nav2_costmap_2d/src/costmap_2d_publisher.cpp

```diff
@@ -181,13 +183,30 @@ void Costmap2DPublisher::prepareCostmap()
   }
 }
 
-void Costmap2DPublisher::publishCostmap()
+std::unique_ptr<nav2_msgs::msg::CostmapUpdate> Costmap2DPublisher::get_raw_costmap_update_msg()
```

Comment:
I was just wondering if you could kindly explain why you use `std::unique_ptr` for the return type.   
Returning the variable with `std::unique_ptr` is unnecessary due to Return Value Optimization (RVO).

Reply:
Publishing the `std::unique_ptr` instead of raw msg is a principle adopted in this project, so that is why I've decided to create msg in unique_ptr directly.
[Change publishers to publish unique ptrs ](https://github.com/ros-planning/navigation2/pull/1673)

I believe that the cause for that is related with `Publisher` class implementation. 
When calling `publish(const T & msg)` additional copy of msg is made when using the intra-process communication
[publisher.hpp:L278](https://github.com/ros2/rclcpp/blob/9c098e544ecf191b7c61f63f7f4fac6f6449cedd/rclcpp/include/rclcpp/publisher.hpp#L278), and in the end you have a copy stored in the unique_ptr and pass the ownership further.
When using ` publish(std::unique_ptr<MessageT, MessageDeleter> msg)` you are omitting this copy. Also please take a look at the comments describing each of the overloaded methods.